alphadia/workflow/optimization.py

```diff
@@ -0,0 +1,684 @@
+                f"✅ {self.parameter_name:<15}: {self.workflow.optimization_manager.__dict__[self.parameter_name]:.4f} <= {self.target_parameter:.4f}",
+                verbosity="progress",
+            )
+            return
+
+        new_parameter = self._propose_new_parameter(
+            precursors_df if self.estimator_group_name == "precursor" else fragments_df
+        )
+        just_converged = self._check_convergence(new_parameter, current_step)
+        self.workflow.optimization_manager.fit({self.parameter_name: new_parameter})
+        self.workflow.optimization_manager.fit(
+            {"classifier_version": self.workflow.fdr_manager.current_version}
+        )
+
+        if just_converged:
+            self.has_converged = True
+            self.reporter.log_string(
+                f"✅ {self.parameter_name:<15}: {self.workflow.optimization_manager.__dict__[self.parameter_name]:.4f} <= {self.target_parameter:.4f}",
+                verbosity="progress",
+            )
+
+        else:
+            self.reporter.log_string(
+                f"❌ {self.parameter_name:<15}: {self.workflow.optimization_manager.__dict__[self.parameter_name]:.4f} > {self.target_parameter:.4f} or insufficient steps taken.",
+                verbosity="progress",
+            )
+
+    def plot(self):
+        """Empty method for consistency with AutomaticOptimizer."""
+        pass
+
+
+class AutomaticRTOptimizer(AutomaticOptimizer):
+    def __init__(
+        self,
+        initial_parameter: float,
+        workflow,
+        reporter: None | reporting.Pipeline | reporting.Backend = None,
+    ):
+        """See base class. Optimizes retention time error."""
+        self.parameter_name = "rt_error"
+        self.estimator_group_name = "precursor"
+        self.estimator_name = "rt"
+        self.feature_name = "precursor_count"
+        super().__init__(initial_parameter, workflow, reporter)
+
+    def _check_convergence(self, current_step: int = -1):
+        """Optimization should stop if continued optimization of the parameter is not improving the TODO feature value.
+        This function checks if the previous rounds of optimization have led to a meaningful improvement in the TODO feature value.
+        It also checks if the current step is greater than the minimum number of steps required for optimization.
+
+        Notes
+        -----
+            Because the check for an increase in TODO feature value requires two previous rounds, the function will also initialize for another round of optimization if there have been fewer than 3 rounds.
+
+        Parameters
+        ----------
+        current_step: int
+            The current step in the optimization process. By default it is set to -1, which prevents the optimizer from converging unless min_steps has been set to 0.
+
+        Returns
+        -------
+        bool
+            True if the convergence conditions are met, False otherwise.
+
+        """
+
+        return (
```

Comment:
(nit)

```python

        min_steps_reached = current_step >= self.workflow.config["calibration"]["min_steps"] - 1
        return (
            min_steps_reached and
            len(self.history_df) > 2
            and self.history_df[self.feature_name].iloc[-1]
            < 1.1 * self.history_df[self.feature_name].iloc[-2]
            and self.history_df[self.feature_name].iloc[-1]
            < 1.1 * self.history_df[self.feature_name].iloc[-3]
            
        )
```
1. easier on the eye (imho)
2. if not `min_steps_reached`, the rest of the calculations is skipped (another fistful of CPU cycles saved ;-))